sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ComplexTypes.scala

```diff
@@ -0,0 +1,131 @@
+        if (idx >= 0 && idx < elems.size) {
+          elems(idx)
+        } else {
+          Cast(Literal(null), ga.dataType)
+        }
+    }
+  }
+}
+
+/**
+* push down operations into [[CreateMap]].
+*/
+object SimplifyCreateMapOps extends Rule[LogicalPlan]{
+  object ComparisonResult extends Enumeration {
+    val PositiveMatch = Value
+    val NegativeMatch = Value
+    val UnDetermined = Value
+  }
+
+  def compareKeys(k1 : Expression, k2 : Expression) : ComparisonResult.Value = {
+    (k1, k2) match {
+      case (x, y) if x.semanticEquals(y) => ComparisonResult.PositiveMatch
+      // make surethis is null safe, especially when datatypes differ
+      // is this even possible?
+      case (_ : Literal, _ : Literal) => ComparisonResult.NegativeMatch
+      case _ => ComparisonResult.UnDetermined
+    }
+  }
+
+  case class ClassifiedEntries(undetermined : Seq[Expression],
+                               nullable : Boolean,
+                               firstPositive : Option[Expression]) {
+    def normalize( k : Expression ) : ClassifiedEntries = this match {
+      /**
+      * when we have undetermined matches that might bproduce a null value,
+      * we can't separate a positive match and use [[Coalesce]] to choose the final result.
+      * so we 'hide' the positive match as an undetermined match.
+      */
+      case ClassifiedEntries( u, true, Some(p)) if u.nonEmpty =>
+        ClassifiedEntries(u ++ Seq(k, p), true, None)
+      case _ => this
+    }
+  }
+
+  def classifyEntries(mapEntries : Seq[(Expression, Expression)],
+                      requestedKey : Expression) : ClassifiedEntries = {
+    val res1 = mapEntries.foldLeft(ClassifiedEntries(Seq.empty, nullable = false, None)) {
+      case (prev @ ClassifiedEntries(_, _, Some(_)), _) => prev
+      case (ClassifiedEntries(prev, nullable, None), (k, v)) =>
+        compareKeys(k, requestedKey) match {
+          case ComparisonResult.UnDetermined =>
+            val vIsNullable = v.nullable
+            val nextNullbale = nullable || vIsNullable
+            ClassifiedEntries(prev ++ Seq(k, v), nullable = nextNullbale, None)
+          case ComparisonResult.NegativeMatch => ClassifiedEntries(prev, nullable, None)
+          case ComparisonResult.PositiveMatch => ClassifiedEntries(prev, nullable, Some(v))
+        }
+    }
+    val res = res1.normalize( requestedKey )
+    res
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan.transformExpressionsUp {
+      // attempt to unfold 'constant' key extraction,
+      // this enables other optimizations to take place.
+      case gmv @ GetMapValue(cm @ CreateMap(elems), key) =>
+        val kvs = cm.keys.zip(cm.values)
+        val classifiedEntries = classifyEntries(kvs, key)
+        classifiedEntries match {
+          case ClassifiedEntries(Seq(), _, None) => Literal.create(null, gmv.dataType)
+          case ClassifiedEntries(`elems`, _, None) => gmv

```

Comment:
@hvanhovell , do you prefer simplifying this like so:
case ClassifiedEntries(es, _, None) if es == elems => gmv